firedrake/petsc.py

```diff
@@ -334,3 +370,71 @@ def garbage_view(obj: Any):
         PETSc.garbage_view(comm)
     else:
         raise FiredrakePETScError("No comm found, cannot view garbage")
+
+
+external_packages = get_external_packages()
+
+# Setup default partitioner
+# Manually define the priority until
+# https://petsc.org/main/src/dm/partitioner/interface/partitioner.c.html#PetscPartitionerGetDefaultType
+# is added to petsc4py
+partitioner_priority = ["parmetis", "ptscotch", "chaco"]
+for partitioner in partitioner_priority:
+    if partitioner in external_packages:
+        DEFAULT_PARTITIONER = partitioner
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(partitioner_priority)
+        + " found, defaulting to PETSc simple partitioner. This may not be optimal."
+    )
+    DEFAULT_PARTITIONER = "simple"
+
+# Setup default direct solver
+direct_solver_priority = ["mumps", "superlu_dist", "pastix"]
+for solver in direct_solver_priority:
+    if solver in external_packages:
+        DEFAULT_DIRECT_SOLVER = solver
+        DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": solver}
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(direct_solver_priority)
+        + " found, defaulting to PETSc LU. This will only work in serial."
+    )
+    DEFAULT_DIRECT_SOLVER = "petsc"
+    DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": "petsc"}
```

Comment:
[connorjward](https://github.com/connorjward) [Mar 13, 2024](https://github.com/firedrakeproject/firedrake/pull/3449#discussion_r1522817604)

I don't really like this pattern. We have constants here that aren't really constants. I would probably prefer this to be called DEFAULT_MUMPS_SOLVER_PARAMETERS. That or make this a function get_default_direct_solver_options("mumps").

I also think that if users are advanced enough to be playing with solver options then we shouldn't obfuscate things behind ambiguous variables. Perhaps this sort of thing could be put into conftest.py as a fixture so we get the right behaviour for the tests but not touch the user experience?